docs/rfcs/0004-agent-flow.md

```diff
@@ -0,0 +1,54 @@
+# Agent Flow - Agent Utilizing Components 
+
+* Date: 2022-03-30
+* Author: Matt Durham (@mattdurham)
+* PR: [grafana/agent#1538](https://github.com/grafana/agent/pull/1538)
+* Status: Draft
+
+## Overarching Problem Statement
+
+The Agents configuration and onboarding is difficult to use. Viewing the effect of configuration changes on telemetry data is difficult. Making the configuration simplier, composable and intuitive allieviates these concerns.
+
+
+## Description
+
+Agent Flow is intended to solve real world needs that the Grafana Agent team have idenfified in conversations with users and developers. 
+
+These broadly include 
+
+- Current configuration is complex and opaque
+- Lack of introspection within the agent
+    - Questions about what telemetry data are being sent
+    - Are rules applying correctly?
+    - How does filtering work?
+- Users have been requesting additional capabilites and adding new features is hard due to coupling between systems, some examples include
+    - Remote Write, Different Input Formats
+    - Different output formats
+    - Filtering ala Relabel Configs is complex and hard to figure out when they occur
+- Lack of understanding how telemetry data moves through agent
+    - Other systems use pipeline/extensions to allow users to understand how data moves through the system
+
+# 1. Introduction and Goals 
+
+This design document describes Agent Flow, which is a system for describing the execution and configuration of telemetry data, along with any secondary configuration required to configure the Grafana Agent. 
+
+Agent Flow refers to both the execution, configuration and visual configurator of data flow.
+
+The primary goals of Agent Flow are
+
+1. Allow users to easily configure the system
+2. Allow users to understand the system
+3. Allow users to inspect the system
+4. Allow developers to easily add components to the system
+
+Goals of this design document
+
+* Define high level concepts and goals
+* Determine whether Agent Flow is a feature that is valuable and useful
+* Define high level concepts of the execution path
+
+This document represents ideals and not technical implementation. 
```

Comment:
nit: I would suggest the design document focus entirely on the core parts of the "what" and "why" from a user's perspective: how will a user configure Grafana Agent Flow? How do they introspect Grafana Agent Flow? Ideally, users shouldn't have to read another design document to understand the core of the proposal. 

I'm saying "core" on purpose: further design documents can layer on top of the core proposal to add further details from a user's perspective. The slimmer we can make the base proposal, the easier it'll be to reach consensus, provided it still has enough detail for users to understand what is being proposed. 

This might have been what you meant by ideals, but I'd like to clarify to see if we're on the same page :)

Reply:
Since the front-end design is in my head, how would you feel if a goal of Agent Flow is the capability to support a GUI. It means we only have to address the supportability of a front end and not the actual implementation.

Reply:
👍 You could generalize that as "flow configs must be easily machine-readable and machine-writable to support tooling around the configs, such as code formatters or a GUI"